Fetch and use fields associated with a parameter to parse parameter values found in URL query

daltojohnso opened the pull request:
Fixes #15498 

We aren't consistently fetching and using `Field` metadata to correctly parse parameter values found in the URL. This PR fixes that.

1. Moves logic from `syncQueryParamWithURL.js` to query param parsing function found in `meta/Parameter.js` so that we parse boolean and numeric values everywhere.
2. In order to have access to the fields associated with a parameter I've moved some rather complex dashboard/selector logic out of the selector function and into `meta/Dashboard.js` so that I can use it directly in an action
3. Parameters in the query builder were initially set by the logic in `syncQueryParamWithURL.js` but I've deleted that code. As a replacement, I'm setting `parameterValues` on the `INITIALIZE_QB` action so that the values get set before queries are run.


Dashboard

https://user-images.githubusercontent.com/13057258/135178960-356a0d49-3987-4cdc-885d-c93aab9c1ba9.mov

Public Dashboard

https://user-images.githubusercontent.com/13057258/135179063-c9c7d333-d19b-401f-8caa-bfdf7bae4af2.mov

(Native) Query Builder

https://user-images.githubusercontent.com/13057258/135178975-ad8e1900-a2f7-4ef4-89fa-7e31db8fcccf.mov

Public Question

https://user-images.githubusercontent.com/13057258/135178987-20dc9f71-6b42-47b2-a43a-7a5872ec4ecd.mov

daltojohnso committed:
move logic in dashboard selectors to meta/Dashboard

daltojohnso committed:
rmv usage of syncQueryParamsWithURL

daltojohnso committed:
add syncQueryParamsWithURL logic to meta/Parameters.js

daltojohnso committed:
Fix public questions

daltojohnso committed:
fix dashboards

daltojohnso committed:
fix query builder

daltojohnso committed:
don't normalize when value is empty string

daltojohnso committed:
rmv unused setMultipleParameterValues fn

daltojohnso commented:
Lots of areas affected by this change, so would appreciate if you took a look at this change, @flamber

flamber reviewed (commented):
Okay, so I have been beating on this for almost two hours, mainly boolean filter stuff.

It's awesome!
It fixes a correctness problem too (only happens with MySQL on `true`), which is actually an old issue, which should have been fixed long ago. But MySQL is funky (MSSQL too), since they don't have a native boolean type, but does tinyint(1) or bit(1) - see issue 18067.

One problem though, when using a nested SQL it breaks the entire dashboard on refresh. These steps are also shown in the video
1. Native query - Sample Dataset (could be any database type) - run query, save as "Q1"
```
select 0::integer as "integer", true::boolean AS "boolean" union all
select 1::integer as "integer", false::boolean AS "boolean" union all
select null as "integer", true::boolean AS "boolean" union all
select -1::integer as "integer", null AS "boolean"
```
2. Explore results - save as "Nested Q1" - add to dashboard
3. On dashboard - create Category filter and connect to "Nested Q1" - save
4. In filter widget input `true` or `false` (see note below) - and do a browser refresh.
```
TypeError: Cannot read properties of undefined (reading 'isNumeric')
    at app-main.bundle.js:2
    at Array.every (<anonymous>)
    at Te (app-main.bundle.js:2)
    at Ne (app-main.bundle.js:2)
    at app-main.bundle.js:2
    at Array.map (<anonymous>)
    at De (app-main.bundle.js:2)
    at app-main.bundle.js:2
    at l (367.bundle.js:2)
    at Generator._invoke (367.bundle.js:2)
(anonymous) @ Dashboard.jsx:154
```

I don't know if the problem comes from the filter input on master is suddenly "plain text", which is actually a big regression in master compared to 0.40.5. I'm creating a separate issue for that.

https://user-images.githubusercontent.com/1447303/135256157-1243de3f-9b41-4375-9b53-8080f18e65b1.mp4

daltojohnso commented:
Thanks for catching that, let me try to fix it.

daltojohnso committed:
replace selector logic with meta/Dashboard fn

daltojohnso committed:
filter out nil fields

we need to do more work before we can properly use
fields from nested questions when parsing parameter values.

daltojohnso committed:
add field to dash parameter so we can correctly type nested native query fields

daltojohnso committed:
fix tests

daltojohnso commented:
OK, @flamber, I think I fixed the bug that you found.

https://user-images.githubusercontent.com/13057258/135547328-bcecc246-1669-4705-bb53-f3e80c761942.mov

flamber reviewed (approved):
@daltojohnso Works beautiful! Love that most of this PR are tests :kissing_heart:

gusaiani reviewed (approved):
I don't feel that my approval has huge weight as a good part of the diff flew over my head.

Just my lack of familiarity with this part of the codebase.

I left a couple of inlined comments.

For the future, seems to me these implementation files that are over 300 or even 700 loc could be broken down.
This will probably allow us to name consts and functions with still-specific but shorter names, and hopefully make things easier to understand.

daltojohnso committed:
code style tweaks per review

daltojohnso commented on frontend/src/metabase/meta/Dashboard.js:
```diff
@@ -268,3 +272,138 @@ export function isDashboardParameterWithoutMapping(parameter, dashboard) {
+        field = rawField && new Field(rawField, metadata);
+      }
+
+      const values = (field && field.fieldValues()) || [];
```
Done

daltojohnso commented on frontend/src/metabase/meta/Dashboard.js:
```diff
@@ -268,3 +272,138 @@ export function isDashboardParameterWithoutMapping(parameter, dashboard) {
+      .uniq()
+      .filter(fieldId => fieldId != null)
+      .value();
+    const fieldIdsWithFKResolved = _.chain(fieldIds)
```
Done

daltojohnso commented on frontend/src/metabase/meta/Dashboard.js:
```diff
@@ -268,3 +272,138 @@ export function isDashboardParameterWithoutMapping(parameter, dashboard) {
   return parameterExistsOnDashboard && !parameterHasMapping;
 }
+
+export function getMappingsByParameter(metadata, dashboard) {
```
I definitely agree with you. I'm spending some time right after this PR to clean this logic up as it is pretty gross.

daltojohnso commented:
@gusaiani -- I don't like putting up PRs this big, but this felt like an exception; the large majority of change here is from unit tests and moving some existing functions around. I might've tweaked ~100 lines of code in this PR. Right after this I'm revisiting the logic in old functions that I've moved around like `getMappingsByParameter` since the code qualities is rather dire.

gusaiani commented:
@daltojohnso 

> I don't like putting up PRs this big, but this felt like an exception; the large majority of change here is from unit tests and moving some existing functions around. I might've tweaked ~100 lines of code in this PR

Absolutely 👍

rlotun commented:
show we backport this so it gets in the 41 release?

daltojohnso commented:
I hesitate because it is a rather complex bug fix to land at the last moment. The fix for https://github.com/metabase/metabase/pull/18223 sort of depends on it, however, so I am wondering if it needs to now.

daltojohnso commented:
@metabase-bot backport release-x.41.x
